### PR TITLE
Chore/wlfe 42

### PR DESCRIPTION
**Pull request by fraincs**
PR for task : Chore/wlfe 42

What was done: 

Set a ThemeProvider component to allow users to control the documentation website theme.

What it does: 

- Check if a theme is stored in localStorage, if not it checks the user's system preference.
- To prevent a flash of color on load, a script is injected before the initialization of the App. This is done in layout.tsx. This is inspired by [Josh Comeau solution](https://www.joshwcomeau.com/react/dark-mode/) for Gatsby.
- Provides a toggle to let the user change the theme independently from it's system settings.
- Upon theme change it saves it to localStorage. This is as soon as the theme is set, meaning that a change of system preference won't change the saved theme until a user clicks the toggle. If deemed necessary the logic could be modified to only save to localStorage on toggleClick.

What it does not do: 

- If a user were to change it's system color scheme while browsing the doc, the website would not react until a refresh. This could be added if deemed necessary.

Known Issues: 

- Since a script is injected before the initialization of the App as stated below, NextJS issues a warning `app-index.js:23 Warning: Extra attributes from the server: ` since the data-theme attribute is added but not present in the SSR generated website. I've yet to find a fix.

### CODE REVIEW

File: apps/docs/app/globals.css [hunk 1]
```diff
@@ -25,14 +23,12 @@ a {
 }
 
 main {
```
Comment (alexasselin008) [inline apps/docs/app/globals.css:25]: Adding spacing for the header directly in the main  style feels weird. 

can't we put the body at 100vh instead?

File: apps/docs/app/layout.tsx [hunk 2]
```diff
@@ -1,22 +1,38 @@
...
-    description: "Generated by create next app"
+    title: "Hopper Documentation",
+    description: "The Hopper Design System Documentation Hub"
 };
 
 export default function RootLayout({
     children
 }: {
     children: React.ReactNode;
 }) {
```
Comment (alexasselin008) [inline apps/docs/app/layout.tsx:15]: should be extracted in RootLayoutProps

Comment (alexasselin008) [reply apps/docs/app/layout.tsx:15]: interface RootLayoutProps { 
children: React.ReactNode;
}

File: apps/docs/app/page.tsx [hunk 3]
```diff
@@ -3,7 +3,7 @@ import { Title } from "@/components/Title/Title";
 export default function Home() {
     return (
         <main>
-            <Title as="h1" level={3}>The first steps of Untitled project</Title>
+            <Title as="h1" level={3}>The first  steps of Hopper</Title>
```
Comment (alexasselin008) [inline apps/docs/app/page.tsx:6]: addded extra space by accident

File: apps/docs/components/Utils/ThemeProvider/ThemeHelper.tsx [hunk 4]
```diff
@@ -0,0 +1,16 @@
+export type ColorScheme = "light" | "dark";
```
Comment (alexasselin008) [inline apps/docs/components/Utils/ThemeProvider/ThemeHelper.tsx:1]: folder Utils should be renamed to utils in lowercase

Comment (alexasselin008) [inline apps/docs/components/Utils/ThemeProvider/ThemeHelper.tsx:1]: rename this file to get-initial-color-mode.ts or inline it in the Theme Provider

Comment (alexasselin008) [reply apps/docs/components/Utils/ThemeProvider/ThemeHelper.tsx:1]: file extension should not be tsx if there is no JSX in it (the <Div> like tags)

File: apps/docs/components/Utils/ThemeProvider/ThemeProvider.tsx [hunk 5]
```diff
@@ -0,0 +1,39 @@
+"use client";
+
+import React from "react";
+import { getInitialColorMode } from "./ThemeHelper";
+
+export type ColorScheme = "light" | "dark";
```
Comment (alexasselin008) [inline apps/docs/components/Utils/ThemeProvider/ThemeProvider.tsx:6]: ColorScheme is already defined in the Theme Helper

File: apps/docs/components/Utils/ThemeProvider/ThemeProvider.tsx [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+interface ColorSchemeContextType {
+    colorMode: ColorScheme;
+    setColorMode?: (newColorScheme: ColorScheme) => void;
```
Comment (alexasselin008) [inline apps/docs/components/Utils/ThemeProvider/ThemeProvider.tsx:10]: i would make setColorMode required nad change the default value to 

```    
setColorMode: () => {}
```

File: apps/docs/components/Utils/ThemeProvider/ThemeSwitch/ThemeSwitch.tsx [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+import { ThemeContext } from "../ThemeProvider";
+import { Icon } from "./ThemeSwitchIcon/ThemeSwitchIcons";
+
+import "./themeSwitch.css";
+
+export type ColorScheme = "light" | "dark";
+
+export default function ThemeSwitch() {
+    const { colorMode, setColorMode } = React.useContext(ThemeContext);
+    const [mounted, setMounted] = React.useState(false);
```
Comment (alexasselin008) [inline apps/docs/components/Utils/ThemeProvider/ThemeSwitch/ThemeSwitch.tsx:12]: avoir un state qui dit si tu as été mounted c'est vraiment une mauvaise pratique, on devrait éviter ça le plus possible

File: apps/docs/components/Utils/ThemeProvider/ThemeSwitch/ThemeSwitch.tsx [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+    React.useEffect(() => {
```
Comment (alexasselin008) [inline apps/docs/components/Utils/ThemeProvider/ThemeSwitch/ThemeSwitch.tsx:14]: Cette logique là devrait être dans le provider, pas dans le toggle

File: apps/docs/components/Utils/ThemeProvider/ThemeSwitch/ThemeSwitch.tsx [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+        colorMode === "dark" && document.documentElement.setAttribute("data-theme", "dark");
```
Comment (alexasselin008) [inline apps/docs/components/Utils/ThemeProvider/ThemeSwitch/ThemeSwitch.tsx:15]: une fois dans le provider ton useEffect pourrait toujours prendre le colorMode quand il change et l'assigné au document
```
useEffect(() => {
    document.documentElement.setAttribute("data-theme", colorMode)
}, [colorMode])
```

File: apps/docs/components/Utils/ThemeProvider/ThemeSwitch/ThemeSwitch.tsx [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+        const theme: ColorScheme = document.documentElement.getAttribute("data-theme") === "dark"
+            ? "light"
+            : "dark";
+
+        document.documentElement.setAttribute("data-theme", theme);
+        setColorMode && setColorMode(theme);
+    };
+
+    return (
+        <>
```
Comment (alexasselin008) [inline apps/docs/components/Utils/ThemeProvider/ThemeSwitch/ThemeSwitch.tsx:29]: fragment pas nécessaire je crois

File: apps/docs/components/Utils/ThemeProvider/ThemeProvider.tsx [hunk 11]
```diff
@@ -0,0 +1,39 @@
+"use client";
```
Comment (alexasselin008) [inline apps/docs/components/Utils/ThemeProvider/ThemeProvider.tsx:1]: c'Est tu vrm nécessaire?

File: apps/docs/tsconfig.json [hunk 12]
```diff
@@ -37,7 +37,7 @@
     "**/*.tsx",
     ".next/types/**/*.ts",
     ".contentlayer/generated"
-  ],
+, "components/Utils/ThemeProvider/ThemeSwitch/ThemeSwitchIcon"  ],
```
[the remark this answers is not in the document]
Comment (fraincs) [reply apps/docs/tsconfig.json:40]: yes, un renaming et une alert VSCode qui a pas bien marcher!

File: apps/docs/components/Utils/ThemeProvider/ThemeProvider.tsx (hunk 6 above)
Comment (fraincs) [reply apps/docs/components/Utils/ThemeProvider/ThemeProvider.tsx:10]: it breaks when I do so

File: apps/docs/components/Utils/ThemeProvider/ThemeProvider.tsx (hunk 11 above)
Comment (fraincs) [reply apps/docs/components/Utils/ThemeProvider/ThemeProvider.tsx:1]: Ce l'à déjà été mais plus maintenant.

File: apps/docs/components/Utils/ThemeProvider/ThemeSwitch/ThemeSwitch.tsx (hunk 10 above)
Comment (fraincs) [reply apps/docs/components/Utils/ThemeProvider/ThemeSwitch/ThemeSwitch.tsx:29]: Ça build pas sans.

File: apps/docs/components/Utils/ThemeProvider/ThemeHelper.tsx (hunk 4 above)
Comment (fraincs) [reply apps/docs/components/Utils/ThemeProvider/ThemeHelper.tsx:1]: inline it in the Theme Provider would create SSR issues

Comment (fraincs) [reply apps/docs/components/Utils/ThemeProvider/ThemeHelper.tsx:1]: > file extension should not be tsx if there is no JSX in it (the

Since it's imported by tsx files they seem to require typing

File: apps/docs/app/globals.css (hunk 1 above)
Comment (alexasselin008) [reply apps/docs/app/globals.css:25]: you should scope your change in the page instead

File: apps/docs/components/Utils/ThemeProvider/ThemeHelper.tsx (hunk 4 above)
Comment (alexasselin008) [reply apps/docs/components/Utils/ThemeProvider/ThemeHelper.tsx:1]: moving the getInitialColorMode inside the ThemePRoviderFile will not cause SSR issues.

For the extension, i meant the file extension should be changed to ThemeHelper.ts. It will not cause issues

File: apps/docs/components/Utils/ThemeProvider/ThemeProvider.tsx (hunk 6 above)
Comment (alexasselin008) [reply apps/docs/components/Utils/ThemeProvider/ThemeProvider.tsx:10]: whats the error? eslint? you can disable the rule for that line

File: apps/docs/components/Utils/ThemeProvider/ThemeSwitch/ThemeSwitch.tsx (hunk 7 above)
Comment (alexasselin008) [reply apps/docs/components/Utils/ThemeProvider/ThemeSwitch/ThemeSwitch.tsx:12]: renomme à [hasReadLocalStorage, setHasReadLocalStorage] pis ça va être correct

File: apps/docs/components/Utils/ThemeProvider/ThemeHelper.tsx (hunk 4 above)
Comment (fraincs) [reply apps/docs/components/Utils/ThemeProvider/ThemeHelper.tsx:1]: > For the extension, i meant the file extension should be changed to ThemeHelper.ts. It will not cause issues
Yes I realized afterward that the x was not necessary :)

File: apps/docs/components/Utils/ThemeProvider/ThemeProvider.tsx (hunk 6 above)
Comment (fraincs) [reply apps/docs/components/Utils/ThemeProvider/ThemeProvider.tsx:10]: > whats the error

Don't use `{}` as a type. `{}` actually means "any non-nullish value".
- If you want a type meaning "any object", you probably want `object` instead.
- If you want a type meaning "any value", you probably want `unknown` instead.
- If you want a type meaning "empty object", you probably want `Record<string, never>` instead.

File: apps/docs/components/Utils/ThemeProvider/ThemeProvider.tsx (hunk 11 above)
Comment (fraincs) [reply apps/docs/components/Utils/ThemeProvider/ThemeProvider.tsx:1]: Nécessaire du au createContext

File: apps/docs/components/Utils/ThemeProvider/ThemeProvider.tsx [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+interface ColorSchemeContextType {
+    colorMode: ColorScheme;
+    setColorMode: (newColorScheme: ColorScheme) => void;
+}
+
+interface ThemeProviderProps {
+    children: React.ReactNode;
+}
+
+export const ThemeContext = React.createContext<ColorSchemeContextType>({
```
Comment (alexasselin008) [inline apps/docs/components/Utils/ThemeProvider/ThemeProvider.tsx:15]: peux tu essayer: 

```
export const ThemeContext = React.createContext<ColorSchemeContextType>({
    colorMode: "light",
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    setColorMode: () => {}
});

export const ThemeProvider = ({ children }: ThemeProviderProps) => {
    const [colorMode, setColorMode] = React.useState(() => getInitialColorMode());

    React.useEffect(() => {
        document.documentElement.setAttribute("data-theme", colorMode);
        window.localStorage.setItem("hdTheme", value);
    }, [colorMode]);

    return (
        <ThemeContext.Provider value={{ colorMode, setColorMode }}>
            {children}
        </ThemeContext.Provider>
    );
};
```

et dans la switch

```
    const toggleTheme = () => {
        const theme: ColorScheme = colorMode === "dark"
            ? "light"
            : "dark";

        setColorMode(theme);
    };
```

File: apps/docs/components/Utils/ThemeProvider/ThemeSwitch/themeSwitch.css [hunk 14]
```diff
@@ -0,0 +1,16 @@
+.hd-themeSwitchButton {
```
Comment (franckgaudin) [inline apps/docs/components/Utils/ThemeProvider/ThemeSwitch/themeSwitch.css:1]: Why did you change the way you name the classes? Based on the BEM methodology. 

Example:
`hd-theme-switch__button` or `hd-theme-switch__btn`

File: apps/docs/components/Utils/ThemeProvider/get-initial-color-mode.ts
Comment (franckgaudin) [inline apps/docs/components/Utils/ThemeProvider/get-initial-color-mode.ts:1]: Use for the same logic everywhere for the naming of files 

`getInitialColorMode.ts`

File: apps/docs/components/Utils/ThemeProvider/ThemeSwitch/ThemeSwitch.tsx (hunk 10 above)
Comment (franckgaudin) [reply apps/docs/components/Utils/ThemeProvider/ThemeSwitch/ThemeSwitch.tsx:22]: A better approach would be to make 

```
    if (colorMode) {
        return (
            <button className="hd-themeSwitchButton"
                aria-label="Toggle Dark Mode"
                onMouseDown={toggleTheme}
            >
                {colorMode === "dark" ? (
                    <Icon icon="sun"/>
                ) : (
                    <Icon icon="moon"/>
                )}
            </button>
        );
    }

    return null;
```

File: apps/docs/app/layout.tsx [hunk 15]
```diff
@@ -1,22 +1,38 @@
...
+    document.documentElement.dataset.theme = getUserPreference();
+  `;
+
     return (
         <html lang="en">
-            <body>
-                <Header />
-                {children}
+            <body id="App">
+                <script dangerouslySetInnerHTML={{ __html: setInitialTheme }} />
```
[the remark this answers is not in the document]
Comment (fraincs) [reply apps/docs/app/layout.tsx:31]: Would this be loaded before NextJS though?

[the remark this answers is not in the document]
Comment (franckgaudin) [reply apps/docs/app/layout.tsx:31]: It loads exactly like the `<script>` tag the difference is that it is optimized for Nextjs.

**Review by franckgaudin — APPROVED** [review]

File: apps/docs/components/Utils/ThemeProvider/getInitialColorMode.ts [hunk 16]
```diff
@@ -0,0 +1,16 @@
+export type ColorScheme = "light" | "dark";
```
Comment (alexasselin008) [inline apps/docs/components/Utils/ThemeProvider/getInitialColorMode.ts:1]: scuse moi ya un dernier trusc dans les fix de review. Ce type là devrait être déclaré dans le ThemePRovider, pas dans la fonction utils qui get la initial color. 


si ça chiale a cause des import, hésite pas a simplement faire import type { ColorScheme } from "./ThemeProvider"

**Review by franckgaudin — APPROVED** [review]

**Review by alexasselin008 — APPROVED** [review]
